### PR TITLE
Remove orphaned NATS pods when upgrading Kyma from 1.X to 2.X

### PR DESCRIPTION
**Pull request by marcobebway**
**Description**

Sometimes nats-operator fails to remove NATS pods during Kyma upgrade from version 1.X to 2.X, which fails the upgrade process.

This pull request introduces a fix in the Eventing component reconciler (as a pre-action step) by removing orphaned NATS pods when upgrading Kyma from version 1.X to 2.X.

> Note: Image bump is done as part of https://github.com/kyma-project/control-plane/pull/1300.

### CODE REVIEW

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go [hunk 1]
```diff
@@ -63,25 +66,36 @@ func (r *removeNatsOperatorStep) Execute(context *service.ActionContext, logger
 	//	return nil
```
Comment (pxsalehi) [inline pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go:66]: I'd say not required for your PR, but if you could remove all these commented out code in these two files, it would be nice! :-)

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go [hunk 2]
```diff
@@ -2,64 +2,80 @@ package preaction
...
...0-destination-rule.yaml\napiVersion: networking.istio.io/v1alpha3\nkind: DestinationRule\nmetadata:\n  name: eventing-nats\nspec:\n  host: eventing-nats.kyma-system.svc.cluster.local\n  trafficPolicy:\n    tls:\n      mode: DISABLE\n---\n# Source: nats/templates/40-cr.yaml\napiVersion: nats.io/v1alpha2\nkind: NatsCluster\nmetadata:\n  name: eventing-nats\nspec:\n  size: 1\n  version: \"2.1.8\"\n  serverImage: \"eu.gcr.io/kyma-project/external/nats\"\n  pod:\n    annotations:\n      sidecar.istio.io/inject: \"false\"\n    labels:\n      helm.sh/chart: nats-1.0.0\n      app: nats\n      app.kubernetes.io/managed-by: Helm\n      kyma-project.io/dashboard: eventing\n    resources:\n      limits:\n        cpu: 20m\n        memory: 64Mi\n      requests:\n        cpu: 5m\n        memory: 16Mi\n  natsConfig:\n    debug: true\n    trace: true\n  template:\n    spec:\n      affinity:\n        podAntiAffinity:\n          preferredDuringSchedulingIgnoredDuringExecution:\n            - podAffinityTerm:\n                labelSelector:\n                  matchLabels:\n                    nats_cluster: eventing-nats\n                topologyKey: kubernetes.io/hostname\n              weight: 100\n"
```
Comment (pxsalehi) [inline pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go:27]: Multiline strings would probably make this way more understandable/editable.

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go [hunk 3]
```diff
@@ -2,64 +2,80 @@ package preaction
...
-		givenStatefulSet: true,
+		givenStatefulSet:    true,
+		givenNatsPodsLength: 0,
 	},
 	{
-		givenStatefulSet: false,
+		givenStatefulSet:    false,
+		givenNatsPodsLength: 3,
 	},
 }
```
Comment (pxsalehi) [inline pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go:46]: Why is `givenNatsPodsLength` > 0 and `givenStatefulSet: true` not a valid test case?

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go [hunk 4]
```diff
@@ -2,64 +2,80 @@ package preaction
...
 	//kyma1xVersion  = "1.24.8"
 	//kyma2xVersion  = "2.0"
+
+	natsPodNameFormat = "nats-%d"
 )
 
 var testCases = []struct {
-	givenStatefulSet bool
+	givenStatefulSet    bool
```
Comment (pxsalehi) [inline pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go:35]: A small comment that `givenStatefulSet` refers to NATS installation via Helm and the NATS pod length is about the cluster size of the NATS operator, would be useful, IMO. (If I have understood this correctly).

**Review by pxsalehi — CHANGES_REQUESTED** [review]

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go (hunk 3 above)
Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go:46]: These two are not dependant on each other. Setting `givenStatefulSet` to `true` or `false` does not affect how the `givenNatsPodsLength` is used in the test.

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go (hunk 1 above)
Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go]: Done

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go (hunk 2 above)
Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: Done

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go (hunk 4 above)
Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: Done

File: pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go (hunk 3 above)
Comment (pxsalehi) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: Yes, I understand. My question is whether we need a test case that represents a cluster where cleanup of NATS pods (created by the operator) were not successful/finished, and there also exists new NATS pods from the statefulset. Can the cluster be in such a state? Or we are sure that until the cleanup is not successful the statefulset is not create?

Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: This state can exist on already broken clusters, and our pre-action step should be able to fix it. But as I mentioned before, the way the test is written does not make `givenStatefulSet` and `givenNatsPodsLength` depend on each other.

Do you think I should add the test anyway?

Comment (pxsalehi) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: If it is valid corner case, why not test it? If I remember correctly, one of the main motivations for this change to the reconciler was to cover these broken cluster cases.

Comment (marcobebway) [reply pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep_test.go]: Done

**Review by pxsalehi — APPROVED** [review]
Thanks!
